[CIR] Add initial support for bit-precise integer types

Lancern opened the pull request:
This PR adds initial support for the bit-precise integer type `_BitInt(N)`. This type goes into the C23 standard, and has already been supported by clang since 2020, previously known as `_ExtInt(N)`.

This PR is quite simple and straight-forward. Basically it leverages the existing `cir.int` type to represent such types. Previously `cir.int` verifies that its width must be a multiple of 8, and this verification has been removed in this PR.

bcardosolopes reviewed (changes requested):
Hi, this is cool, thanks! You say initial support, what's remaining here?

Few extra things this PR needs:
- Add LLVM lowering support, and add tests similar from `clang/test/CodeGen/ext-int.c` and `clang/test/CodeGen/ext-int-cc.c`. I'm open to this being follow up work, but it probably goes well with the size of this PR.
- A bunch of constraints in tablegen assume int types of the primitive ones, so we need to refine those or split them into more variants. Example `CIR_IntType`, `CIR_AnyIntOrFloat`, `IntOrFPPtr`, `IntegerVector` (We either need a better name for them, or change how they are defined). You need to go through each user in CIROps.td and check whether it makes sense for the operations to use anything other than canonical primite types when ints are involved.

bcardosolopes commented:
@sitio-couto you might be interested in this.

Lancern commented:
> You say initial support, what's remaining here?

Well, I'm just not sure whether the changes in this PR is sufficient for a full support, or there are still some unnoticed cases that needs to be resolved. Tests do not follow up that quickly.

> Add LLVM lowering support, and add tests similar from `clang/test/CodeGen/ext-int.c` and `clang/test/CodeGen/ext-int-cc.c`.

Added. Tests that require unimplemented features are not added for now.

> A bunch of constraints in tablegen assume int types of the primitive ones, so we need to refine those or split them into more variants.

Updated.

bcardosolopes commented:
> Well, I'm just not sure whether the changes in this PR is sufficient for a full support, or there are still some unnoticed cases that needs to be resolved. Tests do not follow up that quickly.

Sounds good, just out of curiosity.

> Added. Tests that require unimplemented features are not added for now.

Thanks

bcardosolopes commented on clang/include/clang/CIR/Dialect/IR/CIRTypes.td:
```diff
@@ -139,6 +143,26 @@ class SIntOfWidths<list<int> widths>
+
+def StdUInt : UIntOfWidths<[8, 16, 32, 64]>;
+def StdSInt : SIntOfWidths<[8, 16, 32, 64]>;
+def StdInt : IntOfWidths<[8, 16, 32, 64]>;
```
We already have definitions for a bunch of those, please reuse or merge, but avoid duplicating logic. Example, StdSInt could be:
```
def StdSInt: AnyTypeOf<[SInt8, SInt16, SInt32, SInt64]>;
```
Not tied to a particular approach, but we should think about these things holistically. Also don't add any definiton that isn't used at the moment (e.g. `StdUInt`).

bcardosolopes replied:
There are probably existing ones that aren't used, feel free to clean them up as part of this (if you want)

Lancern replied:
Updated. `UIntOfWidths` and `SIntOfWidths` are not used so I removed them.

bcardosolopes commented on clang/include/clang/CIR/Dialect/IR/CIROps.td:
```diff
@@ -144,7 +144,7 @@ def ObjSizeOp : CIR_Op<"objsize", [Pure]> {
   let arguments = (ins CIR_PointerType:$ptr, SizeInfoType:$kind,
                    UnitAttr:$dynamic);
-  let results = (outs CIR_IntType:$result);
+  let results = (outs StdInt:$result);
```
`Std` is a name that could have so many meanings in C and C++, let's try to avoid it. I suggest we use `PrimitiveInt`, `PrimitiveUInt`, `PrimitiveSInt`.

Lancern replied:
Updated.

bcardosolopes commented on clang/include/clang/CIR/Dialect/IR/CIRTypes.td:
```diff
@@ -53,17 +53,21 @@ def CIR_IntType : CIR_Type<"Int", "int",
     };
+    /// Return true if this is a standard integer type (i.e. not a bit-precise
+    /// integer type).
+    bool isStandard() const {
```
`isPrimitive`

Lancern replied:
Updated.

bcardosolopes commented on clang/include/clang/CIR/Dialect/IR/CIRTypes.td:
```diff
@@ -53,17 +53,21 @@ def CIR_IntType : CIR_Type<"Int", "int",
+    /// Return true if this is a standard integer type (i.e. not a bit-precise
+    /// integer type).
+    bool isStandard() const {
+      return isValidStdIntBitwidth(getWidth());
```
`isValidPrimitiveIntBitwidth`

bcardosolopes replied:
also please update all comments that say "standard"

Lancern replied:
Updated.

bcardosolopes commented on clang/include/clang/CIR/Dialect/IR/CIRTypes.td:
```diff
@@ -375,7 +399,7 @@ def VoidPtr : Type<
 
 // Pointer to int, float or double
-def IntOrFPPtr : Type<
+def StdIntOrFPPtr : Type<
```
`PrimitiveIntOrFPPtr`

Lancern replied:
Updated.

bcardosolopes reviewed (changes requested):
Mostly another round of cosmetic reviews. Please also add aliases to these new ints.

Lancern commented:
Rebased onto the latest `main`.

bcardosolopes commented on clang/lib/CIR/Dialect/IR/CIRDialect.cpp:
```diff
@@ -64,6 +64,10 @@ struct CIROpAsmDialectInterface : public OpAsmDialectInterface {
+      // We only provide alias for standard integer types (i.e. integer types
+      // whose width is divisible by 8).
+      if (intType.getWidth() % 8 != 0)
+        return AliasResult::NoAlias;
```
Not a blocker for this PR, but why not?

bcardosolopes reviewed (approved):
LGTM with one minor comment that can be address in another PR.
